File: clustering/dbscan.py
import matplotlib.pyplot as pl
import numpy as np

class dbscan:
    def __init__(self,n):
        self.NPoints = n
        self.xv = []
        self.yv = []
        self.zv = []
        
        for i in range(n):
            ro = 1
            if np.random.rand() > 0.5:
                ro = ro + 1
            r = ro + 2*(np.random.rand()-0.5)/10

            t = (np.random.rand()-0.5)*2*np.pi
            x = r*np.cos(t)
            y = r*np.sin(t)
            
            self.xv.append(x)
            self.yv.append(y)
            self.zv.append(0)
            
    def neighborhood(self,j):
        N = []
        for i in range(self.NPoints):
            d = np.sqrt((self.xv[j]-self.xv[i])**2 + (self.yv[j]-self.yv[i])**2)
            if (d < 0.5):
                N.append(i)
        
        return N
    
    # Private
    def singleshot(self,i,C):
        visit = []
        visit.append(i)
        while(len(visit) > 0):
            el = visit.pop()
            
            v = self.neighborhood(el)
            if (len(v) > 3 and self.zv[el] == 0):
                self.zv[el] = C
                
                for k in v:
                    visit.append(k)

    def clusterize(self):
        C = 0
        for i in range(self.NPoints):
            if (self.zv[i] == 0):
                v = self.neighborhood(i)
                if (len(v) > 3):
                    C = C + 1
                    self.singleshot(i,C)
                
    def plot(self):
        pl.scatter(self.xv,self.yv,c=self.zv)
                

d = dbscan(300)
d.clusterize()
d.plot()
pl.show()

#def neighborhood(x,y):
    

#NClusters = 0

#for i in range(NPoints):
    #if (z[i] == 0):
        #lista.append((xv[i],yv[i]))
    
    #x,y = lista.pop()
    #V = neighborhood(x,y)

File: unsupervised/clustering/DBSCAN.jl
using PyPlot

function createdata(N)
    R = []
    for i in 1:N
        r = 5+rand()*2
        θ = (rand()-0.5)*π
        x = r*cos(θ)
        y = r*sin(θ)
    
        if (i == 1)
            R = [x y 0]
        else
            R = vcat(R,[x y 0])
        end
        
        x = randn()
        y = randn()
        R = vcat(R,[x y 0])        
    end
    
    return R
end

function Euclidean(x,y)
    return sum((x-y).^2)
end

function neighborhood(p,R,dmin)
    N,M = size(R)
    nei = []
    
    for i in 1:N
        d = Euclidean(R[p,1:2],R[i,1:2])
        if (d < dmin)
            push!(nei,i)
        end
    end
    
    return nei
end

function clusterize(p,R,cluster,Nmin,dmin)
    list = []
    push!(list,p)
    
    while(length(list) != 0)
        q = pop!(list)
        
        nei = neighborhood(q,R,dmin)
        if (R[q,3] == 0 && length(nei) > Nmin)
            R[q,3] = cluster
            while(length(nei) > 0)
                t = pop!(nei)
                push!(list,t)
            end
        end
    end
    
    return R
end    

function Scan(R,Nmin,dmin)
    N,M = size(R)
    
    C = 0
    for i in 1:N
        nei = neighborhood(i,R,dmin)
        if (length(nei) > Nmin && R[i,3] == 0)
            C = C + 1
            R = clusterize(i,R,C,Nmin,dmin)
        end
    end
    
    return R
end

R = createdata(150)
R = Scan(R,3,1)
scatter(R[:,1],R[:,2],10,R[:,3])
show()
